### PR TITLE
expose token cache size in metrics and log

### PR DESCRIPTION
**Pull request by WindzCUHK**
# Description

### log
```console
INFO[2023-12-25T16:35:32+09:00] system_memory_inuse[576.2MB]; go_memstats_heap_alloc_bytes[426.8MB]; accesstoken:cached_token_bytes[218.0MB],entries[200001]; roletoken:cached_token_bytes[176.8MB],entries[200001]; total:cached_token_bytes[394.7MB],entries[400002]; cache_token_ratio:sys[68.2%],heap[92.5%]
```

### metrics
```console
$ curl -sSi http://127.0.0.1:9999/metrics
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 2.120936e+06
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 1.471592e+07
...
# HELP cached_token_bytes Number of bytes cached.
# TYPE cached_token_bytes gauge
cached_token_bytes{type="accesstoken"} 4353
cached_token_bytes{type="roletoken"} 557
# HELP cached_token_entries Number of entries cached.
# TYPE cached_token_entries gauge
cached_token_entries{type="accesstoken"} 5
cached_token_entries{type="roletoken"} 1
```

## minor changes

- github workflow: prevent github release creation on non-tagging events

## Type of change

- [ ] Bug fix
- [x] New feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code

## Related issue/PR

- Closes #70

---

## Checklist

- [x] Followed the guidelines in the CONTRIBUTING document
- [x] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [x] Tested and linted the code
- [x] Commented the code
- [x] Made corresponding changes to the documentation
- [x] Passed all pipeline checking

## Checklist for maintainer

- Use `Squash and merge`
- Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- Delete the branch after merge

### CODE REVIEW

**Comment by mlajkim** [comment]
This is a big change, and I am a bit worried.

**Comment by mlajkim** [comment]
Wonder if it is possible to use the standard JWT, and mimic the response for the Athenz Role Token

**Comment by mlajkim** [comment]
Possibly, should not be included in SIA v3.2.0 release?

File: pkg/token/cache.go [hunk 1]
```diff
@@ -38,19 +42,43 @@ func (k CacheKey) String() string {
...
 	c.cache[k] = t
+
+	// update cache memory usage
+	tokenSize := t.Size()
+	if ok {
+		oldTokenSize := oldToken.Size()
+		c.memoryUsage += int64(tokenSize) - int64(oldTokenSize)
+	} else {
+		keySize := k.Size()
+		c.memoryUsage += int64(keySize + tokenSize)
```
Comment (mlajkim) [inline pkg/token/cache.go:80]:
```suggestion
		c.memoryUsage += int64(k.Size() + tokenSize)
```

Comment (mlajkim) [reply pkg/token/cache.go:80]: or you may change the argument name as `key` not `k`

File: pkg/token/cache.go [hunk 2]
```diff
@@ -38,19 +42,43 @@ func (k CacheKey) String() string {
...
+	return &LockedTokenCache{
+		cache:       make(map[CacheKey]Token),
+		memoryUsage: 0,
+	}
 }
 
 func (c *LockedTokenCache) Store(k CacheKey, t Token) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
+	oldToken, ok := c.cache[k]
```
Comment (mlajkim) [inline pkg/token/cache.go:70]:
```suggestion
	oldToken, exists := c.cache[k]
```

**Review by mlajkim — COMMENTED** [review]
first review

File: pkg/token/cachesize_test.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		// 	wantBucketSize: int64(b) * computeRequiredBuckets(100000),
+		// },
+		{
+			name: "150000 map",
+			args: args{
+				c: make(map[CacheKey]Token, 150000),
+			},
+			wantBucketLen:  32768,
+			wantBucketSize: int64(b) * computeRequiredBuckets(150000),
+		},
```
[the remark this answers is not in the document]
Comment (mlajkim) [reply pkg/token/cachesize_test.go:80]: OK

File: pkg/token/cache.go (hunk 2 above)
Comment (mlajkim) [reply pkg/token/cache.go:70]: (Optional)

Comment (mlajkim) [reply pkg/token/cache.go:70]: go doc uses 

https://go.dev/blog/maps

Comment (WindzCUHK) [reply pkg/token/cache.go:70]: not fix

File: .gitmodules [hunk 4]
```diff
@@ -1,4 +1,5 @@
 [submodule "x509-certificate-exporter"]
-    path = x509-certificate-exporter
-    url = https://github.com/enix/x509-certificate-exporter.git
-    branch = main
+	path = x509-certificate-exporter
+	url = https://github.com/enix/x509-certificate-exporter.git
+	branch = main
+	tag = v3.11.0
```
[the remark this answers is not in the document]
Comment (WindzCUHK) [reply .gitmodules]: reverted in [5d90182](https://github.com/AthenZ/k8s-athenz-sia/pull/69/commits/5d90182ae5f6bb2e35e3915b4ae742cb2dbd19d2)

File: Makefile [hunk 5]
```diff
@@ -88,3 +88,7 @@ submodule-update:
 
 submodule-update-remote:
 	git submodule update --recursive --init --remote
+
+submodule-use-tag:
+	# please update tag in .gitmodules before using this command, tag is not git submodule standard
+	git submodule foreach --recursive 'tag="$$(git config -f $$toplevel/.gitmodules submodule.$$name.tag)";[[ -n $$tag ]] && git reset --hard  $$tag || echo "$$name module has no tag"'
```
[the remark this answers is not in the document]
Comment (WindzCUHK) [reply Makefile]: reverted in 5d90182

File: pkg/token/cachesize_test.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		{
+			name: "10000 map",
+			args: args{
+				c: make(map[CacheKey]Token, 10000),
+			},
+			wantBucketLen:  2048,
+			wantBucketSize: int64(b) * computeRequiredBuckets(10000),
+		},
+		// will fail: https://github.com/golang/go/blob/2184a394777ccc9ce9625932b2ad773e6e626be0/src/runtime/map.go#L346
```
[the remark this answers is not in the document]
Comment (WindzCUHK) [reply pkg/token/cachesize_test.go]: commented in c6caa16

File: pkg/token/cache.go (hunk 1 above)
Comment (WindzCUHK) [reply pkg/token/cache.go]: fixed: 655fb92, ddcb83f

File: pkg/token/cache.go [hunk 7]
```diff
@@ -38,19 +42,41 @@ func (k CacheKey) String() string {
...
 	c.cache[k] = t
+
+	// update cache memory usage
+	tokenSize := t.Size()
+	if ok {
+		c.memoryUsage += int64(tokenSize) - int64(oldToken.Size())
+	} else {
+		c.memoryUsage += int64(k.Size() + tokenSize)
+	}
```
Comment (mlajkim) [inline pkg/token/cache.go:79]:
```suggestion
	// update cache memory usage
	if ok {
		c.memoryUsage += int64(t.Size() - oldToken.Size())
	} else {
		c.memoryUsage += int64(t.Size() + k.Size())
	}
```

File: .gitmodules (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlajkim) [reply .gitmodules]: ok

File: pkg/token/cachesize_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mlajkim) [reply pkg/token/cachesize_test.go]: ok

File: pkg/token/cache.go (hunk 1 above)
Comment (mlajkim) [reply pkg/token/cache.go]: a bit more here:
https://github.com/AthenZ/k8s-athenz-sia/pull/69#discussion_r1437356525

File: pkg/token/cache.go (hunk 7 above)
Comment (mlajkim) [reply pkg/token/cache.go:79]:
```suggestion
	if ok {
		c.memoryUsage += int64(t.Size()) - int64(oldToken.Size())
	} else {
		c.memoryUsage += int64(t.Size() + k.Size())
	}
```

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (mlajkim) [reply Makefile]: ok

**Review by mlajkim — APPROVED** [review]
OK

**Pull request merged**